packages/source-drupal/src/DrupalSource.js

```diff
@@ -0,0 +1,201 @@
+    }
+  }
+
+  constructor(api, options = {}) {
+    if (!options.baseUrl) throw new Error('baseUrl option is required')
+
+    this.options = options
+    this.supportedEntities = {
+      node: Nodes,
+      taxonomy_term: TaxonomyTerms,
+      file: Files,
+      user: Users,
+    }
+    this.entities = {}
+    this.apiSchema = {}
+
+    api.loadSource(store => this.initialize(store))
+  }
+
+  async initialize(store) {
+    try {
+      this.store = store
+      this.apiSchema = await this.fetchJsonApiSchema()
+
+      await this.processEntities()
+    } catch (error) {
+      let { message } = error
+      throw new Error(message)
+    }
+  }
+
+  async fetchJsonApiSchema() {
+    const { baseUrl } = this.options
+
+    if (!baseUrl) throw new Error('baseUrl is required in gridsome.config.js')
+
+    try {
+      let response = await axios.get(`${baseUrl}jsonapi/`)
+
+      /**
+       * response from /jsonapi with axios wrapper is shaped:
+       * data: {
+       *  data: {},
+       *  links: {}
+       * }
+       */
+      let { data: { links } = {} } = response
+
+      /**
+       * normalize /jsonapi reponse
+       * turn this:
+       * {
+       *  'node--article': 'some url',
+       *  'taxonomy_term--category': 'some url',
+       *  ...
+       * }
+       * 
+       * into this:
+       * [
+       *  {
+       *    entityType: 'node',
+       *    entityName: 'article',
+       *    type: 'node--article',
+       *    url: '<jsonapi endpoint>'
+       *  },
+       *  {
+       *    entityType: 'taxonomy_term',
+       *    entityName: 'tags',
+       *    type: 'taxonomy_term--tags',
+       *    url: '<jsonapi endpoint>'
+       *  },
+       * 
+       * ]
+       */
+      return reduce(links, (result, value, key) => {
+        const keySplit = key.split('--') // [entityType, type, url]
+
+        result.push({
+          entityType: keySplit[0],
+          entityName: (keySplit.length > 1) ? keySplit[1] : keySplit[0],
+          type: key,
+          url: value
+        })
+
+        return result
+      }, [])
+    } catch (error) {
+      let { message } = error
+      throw new Error(message)
+    }
+  }
+
+  /**
+   * This method loops over the apiShema created in fetchJsonApiSchema
+   * if entityType (in the apiScheme object) is supported, it creates a new instance
+   * of that supported entity (see the initialize() method in all /src/entities/*)
+   * 
+   * Once each entityType is processed, loop through again and buildRelationships()
+   */
+  async processEntities() {
+    await Promise.all(
+      this.apiSchema.map((api) => {
+        let { entities } = this.options
+        let { entityType, entityName, type, url } = api
+
+        if (this.supportedEntities[entityType] && entities.includes(entityType)) {
+          // creating an instance of the entity class, see ./entities/*
+          this.entities[type] = new this.supportedEntities[entityType](this, { entityType, entityName, type, url })
```

Comment:
```suggestion
          this.entities[type] = new this.supportedEntities[entities[entityType]](this, { entityType, entityName, type, url })
```

Reply:
If we do the mapping that I proposed as a secondary option, then this change won't be necessary. That said, the mapping I proposed is assuming people will more commonly be using the default JSON:API schema.